apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4SamhsaMatcher.java

```diff
@@ -522,43 +524,95 @@ private static String normalizeIcd10Code(String icd10DiagnosisCode) {
-      } else {
-        /*
-         * Fail safe: if we don't know the procedure Coding system, assume the code is
-         * SAMHSA.
-         */
         return true;
       }
     }
-
-    // No blacklisted procedure Codings found: this procedure isn't SAMHSA-related.
     return false;
   }
 
   /**
-   * @param procedureCoding the procedure {@link Coding} to check
-   * @return <code>true</code> if the specified procedure {@link Coding} matches one of the {@link
-   *     #cptCodes} entries, <code>false</code> if it does not
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains at least
+   *     one {@link Coding} and only contains {@link Coding}s that have known coding systems <code>
+   *     false</code> otherwise
+   */
+  private boolean containsOnlyKnownSystems(CodeableConcept procedureConcept) {
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream().map(Coding::getSystem).collect(Collectors.toSet());
+
+    String hcpcsCdSystem = CCWUtils.calculateVariableReferenceUrl(CcwCodebookVariable.HCPCS_CD);
+    String hcpcsSystem = TransformerConstants.CODING_SYSTEM_HCPCS;
+
+    // Valid system url for productOrService coding
+    Set<String> knownHcpcsSystem = Set.of(hcpcsSystem);
+
+    // Additional valid coding system URL for backwards-compatibility
+    // See: https://jira.cms.gov/browse/BFD-1345
+    Set<String> backwardsCompatibleHcpcsSystem = Set.of(hcpcsSystem, hcpcsCdSystem);
+
+    return codingSystems.equals(knownHcpcsSystem)
+        || codingSystems.equals(backwardsCompatibleHcpcsSystem);
+  }
+
+  /**
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains any
+   *     {@link Coding}s that match any of the {@link #cptCodes}, <code>false</code> if they all do
+   *     not
    */
-  private boolean isSamhsaCptCode(Coding procedureCoding) {
+  private boolean isSamhsaCptCode(CodeableConcept procedureConcept) {
     /*
      * Note: CPT codes represent a subset of possible HCPCS codes (but are the only
      * subset that we blacklist from).
      */
-    if (!TransformerConstants.CODING_SYSTEM_HCPCS.equals(procedureCoding.getSystem()))
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream().map(Coding::getSystem).collect(Collectors.toSet());
+
+    if (!codingSystems.contains(TransformerConstants.CODING_SYSTEM_HCPCS)) {
       throw new IllegalArgumentException();
+    }
 
     /*
      * Note: per XXX all codes in icd10DiagnosisCodes are already normalized.
      */
-    return cptCodes.contains(normalizeHcpcsCode(procedureCoding.getCode()));
+    return procedureConcept.getCoding().stream()
+        .anyMatch(
+            procedureCoding -> cptCodes.contains(normalizeHcpcsCode(procedureCoding.getCode())));
```

Comment:
This seems a little disconnected to me, as it checks if there is a HCPCS systems coding, but then checks if _any_ of the codings has a HCPCS code, which theoretically might not be the HCPCS system coding.  Would seem like maybe just checking if it's both a HCPCS system & code would serve the purpose?
If it's not a HCPCS system, this would never have been called in the current implementation.  It's also explicitly checking if a `CodeableConcept` has a HCPCS samhsa code, which would **not** seem to infer a throw if it didn't have a hcpcs system imo, my intuition would expect it to just return false if there is no hcpcs system coding, since it does not contain a hcpcs system code.

Reply:
Indeed, it was a little repetitive. Merged the two checks here and just returned false if we didn't have a HCPCS Coding, which should reduce the redundancy and confusion.